proposals/p2138.md

```diff
@@ -0,0 +1,59 @@
+# Checked and template generic terminology
+
+<!--
+Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+Exceptions. See /LICENSE for license information.
+SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+-->
+
+[Pull request](https://github.com/carbon-language/carbon-lang/pull/2138)
+
+<!-- toc -->
+
+## Table of contents
+
+-   [Abstract](#abstract)
+-   [Problem](#problem)
+-   [Proposal](#proposal)
+-   [Rationale](#rationale)
+
+<!-- tocstop -->
+
+## Abstract
+
+Change terminology from "generics" and "templates" to "checked generics" and
+"template generics". Afterwards, "generics" will be an umbrella term that
+include templates.
+
+## Problem
+
+The C++ community does "generic programming" using templates in C++, and so
+thinks of templates as a mechanism for implementing generics. In Carbon,
+templates and other generic features are more similar than they are different,
+so it is worthwhile to have an easy way to talk about things that apply to both.
+
+## Proposal
+
+Change terminology from "generics" and "templates" to "checked generics" and
+"template generics". Afterwards, "generics" will be an umbrella term that
+include templates.
+
+Some existing design docs give examples of the difference before and after this
+proposal:
+
+-   The [generics overview](/docs/design/generics/overview.md), for example,
+    currently demonstrates the current approach of using "generics" and
+    "templates".
+-   The
+    [generics section of the design overview](/docs/design/README.md#generics)
+    demonstrates the use of the proposed new terminology.
+
+## Rationale
+
+Using the terminology expected by the community supports these goals:
+
+-   [Code that is easy to read, understand, and write](/docs/project/goals.md#code-that-is-easy-to-read-understand-and-write)
+    since our documentation governing the interpretation of that code will be
+    more easily understood and with greater accuracy.
+-   [Interoperability with and migration from existing C++ code](/docs/project/goals.md#interoperability-with-and-migration-from-existing-c-code)
+    since there will be a smaller terminology gap between Carbon and C++.
```

Comment:
I agree with @jonmeow that it'd be good for this proposal to list some considered alternatives. Possibly:

```suggestion
    since there will be a smaller terminology gap between Carbon and C++.

## Alternatives considered

### Status quo: "generic" means "checked"

Use "generic" for checked generics, "template" for template generics, and "template or generic" when we mean both.

Advantages:

-   The terms "generic" and "template" are more concise than "checked generic" and "template generic".
-   More familiar to people coming from C++.

Disadvantages:

-   Does not provide a unifying term for templates and generics, leading to our needing to talk about both in various places, given that they are very similar concepts in Carbon.
-   Does not acknowledge that templates are used for generic programming and as such are a kind of generic.
-   Carbon's template generics aren't exactly the same as C++'s templates, so this may give the wrong intuition for people coming from C++ in some cases.

### Use the name "templates" instead of "template generics"

Use "checked generic" for checked generics, "template" for template generics, and "generic" when we mean both.

Advantages:

-   Shorter term than in this proposal.
-   More familiar to people coming from C++.
-   People will likely use this term anyway rather than saying "template generics".

Disadvantages:

-   Does not parallel the term "unchecked generics".
-   Loses the implication that a template generic is a kind of generic.
-   Carbon's template generics aren't exactly the same as C++'s templates, so this may give the wrong intuition for people coming from C++ in some cases.

### Use the name "unchecked generics" instead of "template generics"

Use "checked generic" for checked generics, "unchecked generic" for template generics, and "generic" when we mean both.

Advantages:

-   Creates a better parallel between "checked generics" and "unchecked generics".
-   Describes the semantics rather than the implementation strategy, especially given that templating / monomorphization is also the implementation strategy we intend to use for unchecked generics.

Disadvantages:

-   Unfamiliar to people coming from C++.
-   Would either need to rename the `template` keyword or pick new non-keyword syntax for it, or have a mismatch between keywords and terminology.
```

Reply:
I made some changes to what you wrote in https://github.com/carbon-language/carbon-lang/pull/2138/commits/cbeae13b6c741c6a50e23d6b68420b795f8a6826